master/internal/rm/multirm/multirm.go

```diff
@@ -0,0 +1,358 @@
+	resolvedRMName := m.getRM(rmName)
+	return m.rms[resolvedRMName].GetExternalJobs(resolvedRMName, rpName)
+}
+
+// GetAgents returns all agents across all resource managers.
+func (m *MultiRMRouter) GetAgents() (*apiv1.GetAgentsResponse, error) {
+	res, err := fanOutRMCall(m, func(rm rm.ResourceManager) (*apiv1.GetAgentsResponse, error) {
+		return rm.GetAgents()
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	all := &apiv1.GetAgentsResponse{}
+	for _, r := range res {
+		all.Agents = append(all.Agents, r.Agents...)
+	}
+	return all, nil
+}
+
+// GetAgent routes a GetAgent request to the specified resource manager & agent.
+func (m *MultiRMRouter) GetAgent(rmName string, req *apiv1.GetAgentRequest) (*apiv1.GetAgentResponse, error) {
+	resolvedRMName := m.getRM(rmName)
+	return m.rms[resolvedRMName].GetAgent(resolvedRMName, req)
+}
+
+// EnableAgent routes an EnableAgent request to the specified resource manager & agent.
+func (m *MultiRMRouter) EnableAgent(rmName string, req *apiv1.EnableAgentRequest) (*apiv1.EnableAgentResponse, error) {
+	resolvedRMName := m.getRM(rmName)
+	return m.rms[resolvedRMName].EnableAgent(resolvedRMName, req)
+}
+
+// DisableAgent routes an DisableAgent request to the specified resource manager & agent.
+func (m *MultiRMRouter) DisableAgent(rmName string, req *apiv1.DisableAgentRequest) (
+	*apiv1.DisableAgentResponse, error,
+) {
+	resolvedRMName := m.getRM(rmName)
+	return m.rms[resolvedRMName].DisableAgent(resolvedRMName, req)
+}
+
+// GetSlots routes an GetSlots request to the specified resource manager & agent.
+func (m *MultiRMRouter) GetSlots(rmName string, req *apiv1.GetSlotsRequest) (*apiv1.GetSlotsResponse, error) {
+	resolvedRMName := m.getRM(rmName)
+	return m.rms[resolvedRMName].GetSlots(resolvedRMName, req)
+}
+
+// GetSlot routes an GetSlot request to the specified resource manager & agent.
+func (m *MultiRMRouter) GetSlot(rmName string, req *apiv1.GetSlotRequest) (*apiv1.GetSlotResponse, error) {
+	resolvedRMName := m.getRM(rmName)
+	return m.rms[resolvedRMName].GetSlot(resolvedRMName, req)
+}
+
+// EnableSlot routes an EnableSlot request to the specified resource manager & agent.
+func (m *MultiRMRouter) EnableSlot(rmName string, req *apiv1.EnableSlotRequest) (*apiv1.EnableSlotResponse, error) {
+	resolvedRMName := m.getRM(rmName)
+	return m.rms[resolvedRMName].EnableSlot(resolvedRMName, req)
+}
+
+// DisableSlot routes an DisableSlot request to the specified resource manager & agent.
+func (m *MultiRMRouter) DisableSlot(rmName string, req *apiv1.DisableSlotRequest) (*apiv1.DisableSlotResponse, error) {
+	resolvedRMName := m.getRM(rmName)
+	return m.rms[resolvedRMName].DisableSlot(resolvedRMName, req)
+}
+
+func (m *MultiRMRouter) getRM(rmName string) string {
+	if rmName == "" {
+		m.syslog.Infof("RM undefined, routing to default manager")
+		return m.defaultRMName
+	}
+
+	_, ok := m.rms[rmName]
+	if !ok {
+		m.syslog.WithError(ErrRMNotDefined(rmName))
+		return m.defaultRMName
```

Comment:
Why fall back to the default when they pass something and we can't find it? I would think we should return an error here instead, there are some calls where I think not hitting the RM you expect would be bad.

Reply:
Okay I can change that, but is it still appropriate to use the defaultRMName if the rmName given is ""?
I.e., suppose we get an API call for "EnableSlot" on rm "" -- should that be applied to defaultRM or none?

Reply:
yeah, i think defaulting "" to "default" is fine but i wouldn't change "my-custom-rm" to "default" if it isn't fine. i would return an error.